ls1intum/Artemis
# `Development`: Improve programming exercise import input validation

### Checklist
#### General
#### Server
#### Changes affecting Programming Exercises

### Motivation and Context
The `templateParticipation` and the `solutionParticipation`, which are both user-controlled, contain the respective `buildPlanId`, which is used in case an imported programming exercise's build plans shall be copied instead of being newly created. By manipulating the `buildPlanId`, it can be possible to:

- Leak other build plans unauthorized.
- Execute unintended URI requests to Bamboo.

### Description
Generate the `buildPlanId` out of `newExercise` instead taking it from `templateParticipation` and `solutionParticipation`, as `newExercise` is not user-controlled, which prevents manipulations.

### Steps for Testing
Prerequisites:
- 1 Instructor
- 1 Student

1. Log in to Artemis as an Instructor
2. Navigate to Course Administration
3. Open any course
4. Navigate to Exercises
5. Create a hello world programming exercise with auxiliary repositories
6. Check on Bamboo if the build plans work correctly (Template fails, but just due to failing test cases; Solution succeeds)
7. Log in to Artemis as a Student
8. Participate in the programming exercise
9. Log in to Artemis as an Instructor
10. Check on Bamboo if the build plan work correctly (Student plan succeeds)
11. Import the previously created programming exercise
12. Check on Bamboo if the build plans work correctly (Template fails, but just due to failing test cases; Solution succeeds)
13. Log in to Artemis as a Student
14. Participate in the programming exercise
15. Log in to Artemis as an Instructor
16. Check if the build plans work correctly

#### Exam Mode Testing
Do the same as above, but this time:
- Import the exercise into an exam
- Import a exercise group into an exam
- Import an exam

In each step, check if the build plans work as expected.

### Review Progress

#### Performance Review
#### Code Review
#### Manual Tests
#### Exam Mode Test

### Test Coverage







### Screenshots
